cmd/aries-agent-mobile/pkg/api/storage.go

```diff
@@ -6,64 +6,95 @@ SPDX-License-Identifier: Apache-2.0
-	// underlying store
-	//
-	// Args:
-	//
-	// startKey: Start of the key range, include in the range.
-	// endKey: End of the key range, not include in the range.
-	//
-	// Returns:
-	//
-	// StoreIterator: iterator for result range
-	Iterator(startKey, endKey string) StoreIterator
-
-	// Delete will delete a record with k key
-	Delete(k string) error
+	// Put stores the key + value pair along with the (optional) tags.
+	// If key is empty or value is nil, then an error will be returned.
+	Put(key string, value []byte, tags ...spi.Tag) error
+
+	// Get fetches the value associated with the given key.
+	// If key cannot be found, then an error wrapping ErrDataNotFound will be returned.
+	// If key is empty, then an error will be returned.
+	Get(key string) ([]byte, error)
+
+	// Get fetches all tags associated with the given key.
+	// If key cannot be found, then an error wrapping ErrDataNotFound will be returned.
+	// If key is empty, then an error will be returned.
+	GetTags(key string) ([]spi.Tag, error)
+
+	// GetBulk fetches the values associated with the given keys.
+	// If no data exists under a given key, then a nil []byte is returned for that value. It is not considered an error.
+	// Depending on the implementation, this method may be faster than calling Get for each key individually.
+	// If any of the given keys are empty, then an error will be returned.
+	GetBulk(keys ...string) ([][]byte, error)
+
+	// Query returns all data that satisfies the expression. Expression format: TagName:TagValue.
+	// If TagValue is not provided, then all data associated with the TagName will be returned.
+	// For now, expression can only be a single tag Name + Value pair.
+	// If no options are provided, then defaults will be used.
+	Query(expression string, options ...spi.QueryOption) (StoreIterator, error)
+
+	// Delete deletes the key + value pair (and all tags) associated with key.
+	// If key is empty, then an error will be returned.
+	Delete(key string) error
+
+	// Batch performs multiple Put and/or Delete operations in order.
+	// Depending on the implementation, this method may be faster than repeated Put and/or Delete calls.
+	// If any of the given keys are empty, then an error will be returned.
+	Batch(operations []spi.Operation) error
+
+	// Flush forces any queued up Put and/or Delete operations to execute.
+	// If the Store implementation doesn't queue up operations, then this method is a no-op.
+	Flush() error
+
+	// Close closes this store object, freeing resources. For persistent store implementations, this does not delete
+	// any data in the underlying databases.
+	Close() error
 }
 
-// StorageProvider defines storage interface.
-type StorageProvider interface {
-	// OpenStore opens a store with given name space and returns the handle
-	OpenStore(name string) (Store, error)
+// StoreIterator allows for iteration over a collection of entries in a store.
+type StoreIterator interface {
```

Comment:
Isn't this defined in spi?

Reply:
My understanding from talking to @soluchok was that this duplication of interfaces is necessary for the mobile bindings. Perhaps @soluchok can elaborate?

Reply:
The main reason to have one more interface was related to the `Release` method. `Release` is a reserved word for ios.
It was not possible to use the original interface. This is why it was wrapped like this `Release() { s.StoreIterator.Free() }`. Now, when we do not have `Release` in the iterator interface it might work. @DRK3 It needs to be checked.

See https://github.com/trustbloc/agent-sdk/pull/49#discussion_r520000674

Reply:
@soluchok Can you check and get back to me? Otherwise I'll have to continue with the PR as-is following the current pattern, since I don't want to break the mobile bindings.

Reply:
@DRK3 you can check it by building mobile builds. If it is possible to inject storage from the iOS and Android then it works. 
How do you know that your code works? Have you built it (for iOS and Android)?
I don’t think it will work `Put(key string, value []byte, tags ...spi.Tag) error` bc of `...spi.Tag`.
Same here `Query(expression string, options ...spi.QueryOption)` bc of `options ...spi.QueryOption`.

Reply:
https://pkg.go.dev/golang.org/x/mobile/cmd/gobind#hdr-Type_restrictions

Reply:
@soluchok How do I build for mobile?

Also, according to the link you sent, it sounds like everything in the new storage package should be supported, including `spi.Tag` and `spi.QueryOption`. So now I'm wondering if we need to have this extra interface at all.

Reply:
I don’t think that this will work `tags ...spi.Tag` and `options ...spi.QueryOption`.

`How do I build for mobile?` - we have a readme https://github.com/hyperledger/aries-framework-go/tree/main/cmd/aries-agent-mobile

Reply:
@soluchok Why won't it work? Do all structs also have to be defined in that file as well? The documentation you linked to seems to suggest that all the types I use are supported.

Reply:
I don’t know how `tags ...spi.Tag` expression can be converted to java (probably it will work somehow). Just make sure your code is compatible. The current code is not compilable.

```
github.com/hyperledger/aries-framework-go/cmd/aries-agent-mobile
gobind/gobind
# gobind/gobind
gobind/go_apimain.go:2797:10: cannot use (*proxyapi_Store)(res_0_ref) (type *proxyapi_Store) as type "github.com/hyperledger/aries-framework-go/cmd/aries-agent-mobile/pkg/api".Store in assignment:
	*proxyapi_Store does not implement "github.com/hyperledger/aries-framework-go/cmd/aries-agent-mobile/pkg/api".Store (missing Batch method)
gobind/go_configmain.go:171:7: cannot use (*proxyapi_StorageProvider)(_v_ref) (type *proxyapi_StorageProvider) as type "github.com/hyperledger/aries-framework-go/cmd/aries-agent-mobile/pkg/api".StorageProvider in assignment:
	*proxyapi_StorageProvider does not implement "github.com/hyperledger/aries-framework-go/cmd/aries-agent-mobile/pkg/api".StorageProvider (missing GetOpenStores method)
gobind/go_storagemain.go:34:7: cannot use (*proxyapi_StorageProvider)(_v_ref) (type *proxyapi_StorageProvider) as type "github.com/hyperledger/aries-framework-go/cmd/aries-agent-mobile/pkg/api".StorageProvider in assignment:
	*proxyapi_StorageProvider does not implement "github.com/hyperledger/aries-framework-go/cmd/aries-agent-mobile/pkg/api".StorageProvider (missing GetOpenStores method)
gobind/go_storagemain.go:84:14: cannot use (*proxyapi_StorageProvider)(_param_p0_ref) (type *proxyapi_StorageProvider) as type "github.com/hyperledger/aries-framework-go/cmd/aries-agent-mobile/pkg/api".StorageProvider in assignment:
	*proxyapi_StorageProvider does not implement "github.com/hyperledger/aries-framework-go/cmd/aries-agent-mobile/pkg/api".StorageProvider (missing GetOpenStores method)
gomobile: go build -v -ldflags -s -w -buildmode=c-shared -o=/var/folders/3l/pr52sz3j3t19m1mqgn7mjyz00000gn/T/gomobile-work-135526515/android/src/main/jniLibs/armeabi-v7a/libgojni.so ./gobind failed: exit status 2

make: *** [bindings-android] Error 1
```

Reply:
@soluchok OK, I will get Android Studio and Xcode set up and try making those bindings.

Reply:
@soluchok I've fixed up the mobile storage interface and wrapper to support the new storage interface, and the bindings seem to be generated correctly now. Can you let me know if the new code looks ok to you?